lib/errors.go

```diff
@@ -0,0 +1,55 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package lib
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common"
+)
+
+// DecoratedError implements error source location rendering.
+type DecoratedError struct {
+	AST *cel.Ast
+	Err error
+}
+
+func (e DecoratedError) Error() string {
+	if e.Err == nil {
+		return "<nil>"
+	}
+	type nodeIDer interface {
+		NodeID() int64
+	}
+	nodeErr, ok := e.Err.(nodeIDer)
```

Comment:
Would it make sense to use `errors.As` here to check the tree in case any wrapped errors have a NodeID?

Reply:
We can't use `errors.As` since it takes a concrete type. We *can* implement our own unwrap.